### ISSUE TITLE
Merge with xeno?

### ISSUE THREAD

**Issue by k0001** [k0001/xmlbf#3]
@ocramz suggested we could upstream this work to `xeno`. Maybe it's a good idea! Thanks @ocramz!

Technically, the only `xeno`-specific thing in `xmlbf` is the `xenoNodeToElement` function, in turn used by `runParser`. Perhaps it makes sense to keep everything but these two functions in `xmlbf`. `xeno` could simply export `runParser` from one of its modules. This way, other XML libraries could export their own `runParser` functions, and they would still work with the `Parser` type.

**Comment by k0001** [comment]
Or, `runParser` in `xmlbf` could take `[Xmlbf.Node]` as input, instead of `ByteString`, and then libraries such as `xeno` would be just responsible for generating this `[Xmlbf.Node]` list. I like this better.

Something else to keep in mind: Having `xeno` depend on `xmlbf` will probably add some transitive dependencies. I can try my best to reduce the number of dependencies. But, if this still is not enough, or if you don't want `xeno` to depend on `xmlbf`, we can introduce a new package called `xeno-xmlbf` or `xmlbf-xeno` that ties the two libraries together.

**Comment by ocramz** [comment]
I'm actually unsure about how to proceed here. Perhaps @chrisdone could share his opinion ?

**Comment by chrisdone** [comment]
Sounds like a good idea! E.g. `Xeno.Serialize` which would export From/To classes and types. 

Given that Xeno's focused on efficiency, I think it's worth benchmarking `Xmlbf` against `Xeno.DOM`. How does it perform? For example, if it's close to `Xeno.DOM` then it's worth it. If it's as slow as e.g. `hexpat` then there'd be no point in using xeno only to wrap it in a slow API. You could just use Hexpat which is safer and better tested. If they're close, then I can help with closing the gap somewhat.

The two places of interest to bench yours against Xeno would be:

* Memory: https://github.com/ocramz/xeno/blob/master/bench/Memory.hs
* Speed: https://github.com/ocramz/xeno/blob/master/bench/Speed.hs

`Xeno.DOM` basically doesn't allocate anything, just a big unpacked vector. Xmlbf in the least creates a tree in mainline memory which the garbage collector will have to deal with.

**Comment by k0001** [comment]
Thanks for your input. 

I ended up doing a couple of things. Here's a summary.

1. The `xmlbf` doesn't deal with raw XML fragments as bytes, instead, it only deals with `xmlbf`'s own `Node`s.

2. There is a library called `xmlbf-xeno` that parses the bytes making up a raw XML fragment into `xmlbf`'s own `Node`s.

3. There is a library called `xmlbf-xmlhtml` doing mostly the same as above, but using `xmlhtml` instead of `xeno`. Mostly to prove the point that that `xmlbf` doesn't really care much about raw XML bytes.

4. Performance is not the most important thing I'm looking for in`xmlblf`, instead, things such as `FromXml`, `ToXml` and `df` are the main motivations of the library. See https://hackage.haskell.org/package/xmlbf-0.1/docs/Xmlbf.html

5. I moved the project to GitLab (I've never used GitLab before, but I'd like to give it a try). I'm closing this repo now.
